src/library/kernels/SpGEMM_EM_kernels.cl

```diff
@@ -0,0 +1,1201 @@
+
+    while (stop >= start)
+    {
+        median = (stop + start) / 2;
+        key_median = s_key[median];
+        
+        if (key_input > key_median)
+            start = median + 1;
+        else if (key_input < key_median)
+            stop = median - 1;
+        else
+        {
+            // atomicAdd is not needed since duplicate is not existed in each input row
+            s_val[median] += val_input;
+            *is_new_col = 0;
+            break;
+        }
+    }
+    //return start;
+}
+
+inline
+void binarysearch_sub(__local int   *s_key,
+                  __local vT *s_val,
+                  int            key_input,
+                  vT          val_input,
+                  int            merged_size)
+{
+    int start = 0;
+    int stop  = merged_size - 1;
+    int median;
+    int key_median;
+
+    while (stop >= start)
+    {
+        median = (stop + start) / 2;
+        key_median = s_key[median];
+
+        if (key_input > key_median)
+            start = median + 1;
+        else if (key_input < key_median)
+            stop = median - 1;
+        else
+        {
+            // atomicAdd is not needed since duplicate is not existed in each input row
+            s_val[median] -= val_input;
+            break;
+        }
+    }
+    //return start;
+}
+
+inline
+void binarysearch_global(__global int   *d_key,
+                  __global vT *d_val,
+                  int            key_input,
+                  vT          val_input,
+                  int            merged_size,
+                  bool          *is_new_col)
+{
+    int start = 0;
+    int stop  = merged_size - 1;
+    int median;
+    int key_median;
+
+    while (stop >= start)
+    {
+        median = (stop + start) / 2;
+        key_median = d_key[median];
+
+        if (key_input > key_median)
+            start = median + 1;
+        else if (key_input < key_median)
+            stop = median - 1;
+        else
+        {
+            // atomicAdd is not needed since duplicate is not existed in each input row
+            d_val[median] += val_input;
+            *is_new_col = 0;
+            break;
+        }
+    }
+    //return start;
+}
+
+inline
+void binarysearch_global_sub(__global int   *d_key,
+                  __global vT *d_val,
+                  int            key_input,
+                  vT          val_input,
+                  int            merged_size)
+{
+    int start = 0;
+    int stop  = merged_size - 1;
+    int median;
+    int key_median;
+
+    while (stop >= start)
+    {
+        median = (stop + start) / 2;
+        key_median = d_key[median];
+
+        if (key_input > key_median)
+            start = median + 1;
+        else if (key_input < key_median)
+            stop = median - 1;
+        else
+        {
+            // atomicAdd is not needed since duplicate is not existed in each input row
+            d_val[median] -= val_input;
+            break;
+        }
+    }
+    //return start;
+}
+
+inline
```

Comment:
There are three different algorithms(Heap -sort, bitonic ESC and Merge-Path) used based on the predicted value of nnZ of output matrix. There purpose is same, find duplicate column indices and compress them. Therefore you might see similar functionality across the different kernels.